### ISSUE TITLE
Make Project config information available to runners

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit-console#748]
The TestCentric GUI needed to be able to list all the configurations available for a project. In general, any GUI runners would probably need this in order to list the available configs so the user could select one. It also needed to be able to recognize projects in the first place in order to know when such a menu should be displayed. While this could be done in other ways, the simplest thing was to call the `ProjectService` taking advantage of whatever project loader extensions were installed in the engine.

Because `IProjectService` was not originally provided in the API assembly, I added it there originally. Now that TestCentric is using the packaged NUnit API assembly, it is no longer available. The features that made use of it have now been disabled. Fortunately, they were never released, so it causes no problems to the users right now.

While there are other approaches, such as saving the information in the `TestPackage`, exposing `IProjectService` in the API assembly seems like the simplest thing all around. If this seems like a reasonable thing to do I can provide a PR. I've marked it high priority from my perspective of needing the interface. :smile: 

Please comment @nunit/engine-team !

**Comment by ChrisMaddock** [comment]
Sounds sensible to me. Could you propose an API, Charlie?

**Comment by CharliePoole** [comment]
@ChrisMaddock All I did in my (old) testcentric code was move the current api definition out of the internal namespace in `nunit.engine` and into `nunit.engine.api`. But you're right - this may actually be an opportunity to redefine the API if we need to. I'll take a look at that.

**Comment by CharliePoole** [comment]
Here's the API I was using. It matches NUnit's, except for the `LoadFrom` method, which I added.

The `IProject` interface returned from `LoadFrom` is already in the API assembly because it is used by extenders. In fact, it's a bit odd that I never made `IProjectService` public. I think that was because I wanted runners to not have to worry about projects. In fact, they don't unless they want some internal info like what configurations are supported.

```
    /// <summary>
    /// The IProjectService interface is implemented by ProjectService.
    /// It knows how to load projects in a specific format and can expand
    /// TestPackages based on projects.
    /// </summary>
    public interface IProjectService
    {
        /// <summary>
        /// Returns true if the file indicated is one that this
        /// loader knows how to load.
        /// </summary>
        /// <param name="path">The path of the project file</param>
        /// <returns>True if the loader knows how to load this file, otherwise false</returns>
        bool CanLoadFrom(string path);

        /// <summary>
        /// Loads a project of a known format.
        /// </summary>
        /// <param name="path">The path of the project file</param>
        /// <returns>An IProject interface to the loaded project or null if the project cannot be loaded</returns>
        IProject LoadFrom(string path);

        /// <summary>
        /// Expands a TestPackage based on a known project format, populating it
        /// with the project contents and any settings the project provides. 
        /// Note that the package file path must be checked to ensure that it is
        /// a known project format before calling this method.
        /// </summary>
        /// <param name="package">The TestPackage to be expanded</param>
        void ExpandProjectPackage(TestPackage package);
    }
```

I think there are two decisions to be made here.

1. Should we make the interface public? That seems like an easy decision.

2. Should we add the `LoadFrom` method or do something else? Right now, the only thing I need is the names of the configs, so we could be minimalist and have a method that does just that. It's the choice I'd want to make if `IProject` were not already defined publicly. Since it is, however, I think just returning the entire project definition is the best course. That way we don't have to predict what other bits might be needed from it.

@nunit/engine-team What do you think?

**Comment by ChrisMaddock** [comment]
Hi Charlie,

Can I add an option 3? I don't think it's necessary to expose all of this - personally I'd prefer to just take the approach of exposing what we need, and then adding further methods later. 

I would agree with just exposing `IProject LoadFrom(string path)` - but I'm not a big fan of that method signature in general. I'd prefer to see a `bool TryLoadFrom(string path, out IProject project)`, which would be more explicit to the end user about what they're getting - without needing to read the docs about whether to expect null's or exceptions.

Last thought - are we happy that this requires a runner to implement logic to check if a file is a dll or exe, and assume it's a project if not? I'm not sure I have a nice solution to that...but that sort of feels like internal knowledge, and it could be nice for a runner to just pass the engine a file, and get back a list of any config's that are relevant, regardless of the file.

How will you implement this in the TestCentric GUI, I presume by checking the file extension?

**Comment by ChrisMaddock** [comment]
Ooh, option 4...how about the Project service instead exposes known project types? That means a runner can check known project types and `TryLoad` if necessary. It also means that 'anything else' would just go into the test package and fail - but means a runner doesn't need knowledge of the "projects might be anything except dll's and exe's" logic.

**Comment by CharliePoole** [comment]
Two considerations seem to be behind our different views:

1. I was seeing `IProject` (`IProjectLoader` as well) as already exposed. They need to be so that people can write loader extensions. Aside from what I (or others) need to access, it seems inconsistent that `IProjectService` is not exposed as well.

2. I thought we needed to keep `IProjectService` backward-compatible. Thinking about it, I see that could be debatable. On the one hand, it needs to stay backward-compatible __after__ we expose it. I would be willing to make changes to it and wouldn't consider them breaking. HOWEVER, that's me. The project has appeared to move in a direction where changing anything that a user __might__ have accessed is viewed as breaking. Currently, `IProjectService` is defined as a public interface in the engine assembly. Somebody could have referenced it. I would be willing to change it __and__ move it to the API assembly. Would you? What about the rest of the team? [This is why I advocate a written definition of breaking.]

Beyond that... considering the options you suggested...

"Option 3"

I felt exposing `LoadFrom`, which throws if the path given is not a supported project, was OK because we also have `CanLoadFrom`, which the user should have called first. If I were starting from scratch, I'd probably call the latter something like `IsSupportedProject`. But in any case, the runner doesn't need to check for ".dll" or ".exe". It should be asking the service first. Even if the caller "knows" that a particular extension is a project type, e.g. `.nunit`, it doesn't know if the proper extension is installed. Checking for installation is more complicated than just asking.

This could be a style difference as well. I dislike `out` parameters. Also, I'm not a big fan of users who don't read the docs. :smile: 

"Option 4"

Returning a list of known types is a nice addition.

"Option 5" 

We __could__ leave the service interface alone and introduce an entirely new interface that's public. That's what I did for available runtimes, for example. Back in 2017, I added `IAvailbleRuntimes` to the API and let the `RuntimeFrameworkService` implement it. `IRuntimeFrameworkService` didn't get published at all. The change was actually trivial to make.

Long story short...

I'd like to do a PR for this with a published interface that does __not__ replace `IProjectService` at all, just exposes some aspects of it that are needed. I don't have a name for the interface yet, so I'd welcome any suggestions.

**Comment by ChrisMaddock** [comment]
> wo considerations seem to be behind our different views:

> I was seeing IProject (IProjectLoader as well) as already exposed. They need to be so that people can write loader extensions. Aside from what I (or others) need to access, it seems inconsistent that IProjectService is not exposed as well.

I agree that IProject/IProjectLoader are already exposed. I disagree this is inconsistent - the former are necesary to write project loader extensions, the latter has a separate purpose.

> I thought we needed to keep IProjectService backward-compatible. Thinking about it, I see that could be debatable. On the one hand, it needs to stay backward-compatible after we expose it. I would be willing to make changes to it and wouldn't consider them breaking. HOWEVER, that's me. The project has appeared to move in a direction where changing anything that a user might have accessed is viewed as breaking. Currently, IProjectService is defined as a public interface in the engine assembly. Somebody could have referenced it. I would be willing to change it and move it to the API assembly. Would you? What about the rest of the team? [This is why I advocate a written definition of breaking.]

I'm currently working on the basis that the only supported way to integrate with the engine is via nunit.engine.api.dll. Therefore, I think we should currently be able to change IProjectService freely.

This is a big thing I'd like to clear up and lockdown entirely in v4. 😄 

> Beyond that... considering the options you suggested...

> "Option 3"

> I felt exposing LoadFrom, which throws if the path given is not a supported project, was OK because we also have CanLoadFrom, which the user should have called first. If I were starting from scratch, I'd probably call the latter something like IsSupportedProject. But in any case, the runner doesn't need to check for ".dll" or ".exe". It should be asking the service first. Even if the caller "knows" that a particular extension is a project type, e.g. .nunit, it doesn't know if the proper extension is installed. Checking for installation is more complicated than just asking.

> This could be a style difference as well. I dislike out parameters. Also, I'm not a big fan of users who don't read the docs. 😄

I think you're right, this is a style difference! 😛 I'm glad we agree the runner doesn't need to know about dll/exe. 

> "Option 4"

> Returning a list of known types is a nice addition.

> "Option 5"

> We could leave the service interface alone and introduce an entirely new interface that's public. That's what I did for available runtimes, for example. Back in 2017, I added IAvailbleRuntimes to the API and let the RuntimeFrameworkService implement it. IRuntimeFrameworkService didn't get published at all. The change was actually trivial to make.

> Long story short...

> I'd like to do a PR for this with a published interface that does not replace IProjectService at all, just exposes some aspects of it that are needed. I don't have a name for the interface yet, so I'd welcome any suggestions.

I didn't realise we already had an IProjectService in nunit.engine when I started this. 🙂 I agree we want a new interface - but I think we should call the API version IProjectService, and rename the 'internal' version. I think `IProjectService` is the best-fitting name, so should be used for the public interface.

**Comment by CharliePoole** [comment]
@ChrisMaddock OK, I'll do a PR on the basis of your last paragraph. :smile: I think I can work on this without running into conflicts with any pending changes i have.

**Comment by CharliePoole** [comment]
@ChrisMaddock This discussion got me thinking so I have come up with an alternative that doesn't require a new interface. Both approaches are simple enough that I'm doing two PRs and we can pick one.

**Comment by CharliePoole** [comment]
I created two quick and dirty draft PRs, showing two approaches.

Option A - add new public `IProjectService` interface as @ChrisMaddock and I have been discussing at length. While I no longer prefer this option, it got me thinking about what I __actually__ need in the GUI, which BTW I believe is typical for any runner that stays open over time and allows the user to change settings and perform multiple runs. This is PR #753 

Option B - have the existing `ProjectService` add the required information to the package it creates. Interfaces all remain the same. This my preferred option.

Discussion:

Both options perform the necessary work to collect config info in the `ProjectService`. That's important because we don't want existing project extensions to require updates.

In my own GUI, at least, I need ActiveConfig and a list of Config names. They are used to provide a menu of available configs, highlighting the one that is currently Active - the one you get if you don't specify a config. Since the command-line supports multiple projects, I need to determine this on a per-project basis. Most likely, if we used option A, I would end up adding the info to the test packages in the GUI. Having the info there in the first place makes life simpler.

The existing means of communication between the projet service and the runners is through the `TestPackage`. Option B makes use of what is already being done.

As can be seen in the two PRs, option A is simple. Option B is extra-simple. :smile:

Option A makes an arguably trivial breaking change. Option B doesn't.

My thought is to give anyone who wants a short period to comment and then proceed with Option B. What do you think? :smiley:

**Comment by CharliePoole** [comment]
Changed the title of this issue as well. New title more accurately states what is needed. @ChrisMaddock Thanks for all the discussion, which got me here!

**Comment by ChrisMaddock** [comment]
Fixed in #754 I believe
